server/channels/app/session.go

```diff
@@ -27,6 +27,17 @@ func (a *App) CreateSession(c request.CTX, session *model.Session) (*model.Sessi
 		return nil, appErr
 	}
 
+	// remote/synthetic users cannot create sessions. This lookup will already be cached.
+	// Some unit tests rely on sessions being created for users that don't exist, therefore
+	// missing users are allowed.
+	user, appErr := a.GetUser(session.UserId)
+	if appErr != nil && appErr.StatusCode != http.StatusNotFound {
+		return nil, appErr
+	}
```

Comment:
Would it be possible to make this additional check only if shared channels are enabled?

/cc @agarciamontoro @agnivade in case you have performance considerations.

Reply:
@streamer45 I don't think that would satisfy the security requirements.  If Shared Channels was once enabled, and remote/synthetic users were created, we must ensure those accounts cannot be used locally even when Shared Channels is disabled.

Reply:
How would remote users be able to create sessions if the functionality is disabled? I guess we are not checking upon login?

Reply:
From a performance perspective, my take is that session creation happens for every user, but does not happen all that often per user.  The main over-head here is the GetUser lookup, which is also done a few items up the call stack in the Rest API layer so should be cached.

Reply:
Yeah, I guess I am just wondering whether we could simply chat at that point when we have the user object. Not sure if it would overcomplicate though.

Reply:
> How would remote users be able to create sessions if the functionality is disabled? 

Remote users are just entries in the users table that happen to have the remoteID field non-blank. We cannot let them log in regardless of whether Share Channels is enabled or not (this also holds true for MS Teams integration) so checking if the service is enabled doesn't help.  

> I guess we are not checking upon login?

The security issue is that there are many code paths for logging in and there is concern that trying to put checks in all of them is fragile.  All the code paths end up creating a session, which is why I put the check here.

Reply:
@streamer45 I didn't realize auto-merge had been added to the PR.  Either way, I want to make sure your concerns are addressed.  I can revert if needed.

Reply:
That was unexpected. :p I have no blocking concerns. I think the above makes sense. Maybe we could have found a way to pass the user object through that method to avoid fetching again, but I suppose that if everything is cached correctly (as it sounds), we should be okay.